middle_end/inline_and_simplify.ml

```diff
@@ -1307,28 +1311,35 @@ and simplify env r (tree : Flambda.t) : Flambda.t * R.t =
           | Some l ->
             let approx = R.approx r in
             let l, r = simplify env r l in
-            Some l, R.set_approx r (A.meet (R.approx r) approx)
+            Some l,
+            R.set_approx r (A.meet ~really_import_approx (R.approx r) approx)
         in
         let sw = { sw with failaction; consts; blocks; } in
         Switch (arg, sw), r
       end)
   | String_switch (arg, sw, def) ->
     simplify_free_variable env arg ~f:(fun env arg _arg_approx ->
+      let really_import_approx = E.really_import_approx env in
+      let env = E.inside_branch env in
       let sw, r =
         List.fold_right (fun (str, lam) (sw, r) ->
+            let approx = R.approx r in
             let lam, r = simplify env r lam in
-            (str, lam)::sw, r)
+            (str, lam)::sw,
+            R.set_approx r (A.meet ~really_import_approx (R.approx r) approx))
```

Comment:
What about adding `R.meet_approx`?  It might simplify the code in these parts a bit.